### ISSUE TITLE
Major bug (renders esm-tools v4 in their current state **UNUSABLE**): Lack of mandatory files (restart, forcing) does not kill the simulation

### ISSUE THREAD

**Issue by christian-stepanek** [esm-tools/esm_tools#63]
**Description of the problem**
A master student has run a simulation with esm-tools v4. A restart file was not found. The expected behavior is that the simulation run, that discovers lack of a mandatory file, exits with a meaningful message, but does *not* resubmit itself. Yet, in practice, re-submission independently of an error check is exactly what currently happens.

After esm-tools discover a missing mandatory file, the current run iteration exits and resubmits the following run year. The tools do this happily and merrily until exhaustion of the disk quota (related to the "cleanup-after-a-run" bug #55) puts a sad end to this crazy behavior.

Examples:
log file of simulation year 2166
```
Module for cmake version 3.13.3 loaded
Module for udunits version 2.2.26 loaded
Module for grib_api version 1.28.0 loaded.
Module for Intel Compiler and Libraries, ifort, icc, icpc; mkl, ipp, tbb,... version 2018.5.274 loaded.
Module for hdf5 version 1.10.2_gnu loaded
Module for centoslibs version 7.5 loaded.
Module for cdo version 1.9.6 loaded
Please add >module load centoslibs<, if you are on a compute node to get all libraries.
Module for nco version 4.7.7 loaded
Please add >module load centoslibs<, if you are on a compute node to get all libraries.
Module for netcdf version 4.6.2_intel_18.0.5 loaded
Module for Python 3.7 from Intel Parallel Studio 2020 loaded
Module for automake version 1.16.2 loaded
Currently Loaded Modulefiles:
 1) cmake/3.13.3                     7) cdo/1.9.6(default)       
 2) udunits/2.2.26                   8) nco/4.7.7(default)       
 3) gribapi/1.28.0                   9) netcdf/4.6.2_intel       
 4) intel.compiler/18.0.5(default)  10) python3/3.7.7_intel2020  
 5) hdf5/1.10.2_gnu                 11) automake/1.16.2          
 6) centoslibs/7.5(default)         
Module for impi,  version 2018.4.274 loaded.
432:  MPI has been initialized
432:  Running on          288  PEs
  0:  redirecting console output of echam process 0 to files "echam.stdout" and "echa
  0:  m.stderr"
432:  time step size is set to    1800.00000000000      sec
  2: FATAL ERROR in : restart file not available !
  2:  FINISH called from PE: 2
  2: 
  2: --------------------------------------------------------------------------------
  2: 
  2: 
  4: FATAL ERROR in : restart file not available !
  4:  FINISH called from PE: 4
 12: FATAL ERROR in : restart file not available !
  1: FATAL ERROR in : restart file not available !
 12:  FINISH called from PE: 12
 12: 
 12: --------------------------------------------------------------------------------
 12: 
 12: 
  1:  FINISH called from PE: 1
  4: 
[...]
restart/jsbach//restart_PI_ctrl_AWI-ESM-2-1_veg.nc
--  /work/ollie/abose/esm-experiments//PI_ctrl_AWI-ESM-2-1/run_21670101-21671231/work//restart_PI_ctrl_AWI-ESM-2-1_land.nc: 
        --> /work/ollie/abose/esm-experiments//PI_ctrl_AWI-ESM-2-1/run_21670101-21671231/restart/jsbach//restart_PI_ctrl_AWI-ESM-2-1_land.nc
========================================================
database_entry
--------------
submit
------

 + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + 
Submitting sad jobscript to batch system...
cd /work/ollie/abose/esm-experiments//PI_ctrl_AWI-ESM-2-1/run_21670101-21671231/scripts/; sbatch /work/ollie/abose/esm-experiments//PI_ctrl_AWI-ESM-2-1/run_21670101-21671231/scripts//PI_ctrl_AWI-ESM-2-1_compute_21670101-21671231.sad

 + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + 
Exiting entire Python process!
```

logfile of simulation year 2376
```
Module for cmake version 3.13.3 loaded
Module for udunits version 2.2.26 loaded
Module for grib_api version 1.28.0 loaded.
Module for Intel Compiler and Libraries, ifort, icc, icpc; mkl, ipp, tbb,... version 2018.5.274 loaded.
Module for hdf5 version 1.10.2_gnu loaded
Module for centoslibs version 7.5 loaded.
Module for cdo version 1.9.6 loaded
Please add >module load centoslibs<, if you are on a compute node to get all libraries.
Module for nco version 4.7.7 loaded
Please add >module load centoslibs<, if you are on a compute node to get all libraries.
Module for netcdf version 4.6.2_intel_18.0.5 loaded
Module for Python 3.7 from Intel Parallel Studio 2020 loaded
Module for automake version 1.16.2 loaded
Currently Loaded Modulefiles:
 1) cmake/3.13.3                     7) cdo/1.9.6(default)       
 2) udunits/2.2.26                   8) nco/4.7.7(default)       
 3) gribapi/1.28.0                   9) netcdf/4.6.2_intel       
 4) intel.compiler/18.0.5(default)  10) python3/3.7.7_intel2020  
 5) hdf5/1.10.2_gnu                 11) automake/1.16.2          
 6) centoslibs/7.5(default)         
Module for impi,  version 2018.4.274 loaded.
  0:  redirecting console output of echam process 0 to files "echam.stdout" and "echa
  0:  m.stderr"
432:  MPI has been initialized
432:  Running on          288  PEs
432:  time step size is set to    1800.00000000000      sec
  2: FATAL ERROR in : restart file not available !
  2:  FINISH called from PE: 2
  2: 
  2: --------------------------------------------------------------------------------
  2: 
  2: 
  1: FATAL ERROR in : restart file not available !
  1:  FINISH called from PE: 1
  1: 
  1: --------------------------------------------------------------------------------
  1: 
  1: 
  3: FATAL ERROR in : restart file not available !
  3:  FINISH called from PE: 3
  3: 
  3: --------------------------------------------------------------------------------
  3: 
  3: 
  4: FATAL ERROR in : restart file not available !
  4:  FINISH called from PE: 4
  4: 
  4: -------
[...]
========================================================
database_entry
--------------
submit
------

 + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + 
Submitting sad jobscript to batch system...
cd /work/ollie/abose/esm-experiments//PI_ctrl_AWI-ESM-2-1/run_23770101-23771231/scripts/; sbatch /work/ollie/abose/esm-experiments//PI_ctrl_AWI-ESM-2-1/run_23770101-23771231/scripts//PI_ctrl_AWI-ESM-2-1_compute_23770101-23771231.sad

 + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + 
Exiting entire Python process!
```
So, in the current case the esm-tools managed to ignore a missing restart file and pretended to do meaningful simulations for 213 years, until the quota put an end to this rogue behavior.

**Suggested solution**
Somewhere in the new version of the tools there is lack of a check that makes sure that resubmission only occurs if the current run iteration ended without any obvious problem. I am not sure whether the insane "I do not care whether the run works, I will just resubmit anyway"-behavior may be linked to the new feature of the esm-tools to check for known issues (e.g. windspeed bug), to fix these, and then to automatically resubmit a fixed version of the run. This is just a guess on my side, but could be looked at.

**employed run script (on Ollie):**
/home/ollie/stepanek/diagnostic_files_esm_tools_restart_bug/PI_ctrl_AWI-ESM-2-1.yaml

**example runfolder kept for diagnostics:**
/home/ollie/stepanek/diagnostic_files_esm_tools_restart_bug/run_23750101-23751231/

**employed esm_tools versions:**
esm_archiving : unknown version!
esm_autotests : unknown version!
esm_calendar : 4.0.1
esm_database : 4.0.0
esm_environment : 4.0.1
esm_master : 4.0.2
esm_parser : 4.0.2
esm_profile : 4.0.0
esm_rcfile : 4.0.0
esm_runscripts : 4.0.3
esm_tools : 4.0.9
esm_plugin_manager : 4.0.1
esm_version_checker : 4.0.2

**Comment by pgierz** [comment]
One suggestion would be to add the error that is thrown:

"FATAL ERROR in : restart file not available !"

Into the automatic error checking done by ECHAM...

**Comment by christian-stepanek** [comment]
Should this not work the other way around?

If there is any error, then the simulation should not resubmit. Only in specific cases, when there is a tested robust way to fix the error automatically, a resubmission should occur.

Pseudo code
```
submit run_0001
last_error=$?

case $last_error in
127*)
 % known error for windspeed bug 
 Message="We know what to do to fix the problem and resubmit."
 resubmit run_0001
  ;;
130*)
  % some other known error that can be automatically fixed
  Message="We know what to do to fix the problem and resubmit."
  resubmit run_0001
  ;;
*)
  Message="Something else happens and it is not clear how to fix this automatically"
  kill run_0001
  ;;
esac
```
Note, that PYTHON does NOT support CASE-STATEMENTS! This is indeed a very nice feature for a high level language ... [sarcasm]

**Comment by pgierz** [comment]
what does slurm spit back? there's a small shell wrapper that gets fired off. This also hooks up the "resubmit" process. We could just stop that if the model doesn't exit cleanly.

**Comment by pgierz** [comment]
OK, I know how to fix this after a brief Skype. I'll send a fix and instructions how to use it

**Comment by pgierz** [comment]
Fixed in version v4.1.3 of esm-runscripts.

Christian, please run `esm_versions upgrade`, and let me know if it works...
